Lib/csharp/std_map.i

```diff
@@ -236,7 +236,11 @@
       }
 
       void setitem(const key_type& key, const mapped_type& x) {
+%#if __cpp_lib_map_try_emplace >= 201411L
```

Comment:
Testing like this can trigger compiler warnings when the macro isn't defined, e.g.:

```
$ printf '#if __cpp_lib_map_try_emplace >= 201411L\n#endif\n'|g++ -xc++ -std=c++98 -Wundef -c -
<stdin>:1:5: warning: "__cpp_lib_map_try_emplace" is not defined, evaluates to 0 [-Wundef]
```

There isn't an earlier version for `__cpp_lib_map_try_emplace` so I'd suggest just checking if it's defined (i.e. `#ifdef __cpp_lib_map_try_emplace`).

Reply:
Interesting, this code is exactly how cppreference instruct how to test for library features.

I guess you are right, there is no older version of that api so only testing for the macro being defined is enough (we could even test both actually, but I'll leave this for the next version of the api)

Reply:
If the example you're looking at is this one from https://en.cppreference.com/w/cpp/feature_test:

```
#if __cpp_constexpr >= 201304                // Check for a specific version of a feature
#  define CONSTEXPR constexpr
#else
#  define CONSTEXPR inline
#endif
```

then the reason for that is that there are multiple possible different values of `__cpp_constexpr` corresponding to various new `constexpr` features - the first was `200704L` but this is showing testing for a later version.  That's why I particularly noted that there isn't an earlier version of `__cpp_lib_map_try_emplace`.

A more similar example to ours is:

```
#if __cpp_binary_literals                    // Check for the presence of a feature
    unsigned mask1 = 0b11000000;
    unsigned mask2 = 0b00000111;
#else
    unsigned mask1 = 0xC0;
    unsigned mask2 = 0x07;
#endif
```

That also emits warnings with `-Wundef` with GCC and clang though.  I don't really see why you'd prefer `#if` to `'#ifdef` for that one; I can see that checking defined-ness as well as the value could be viewed as needlessly uglier if you control the warning flags in use, but for a context like SWIG's generated code or public library headers you'd want to use something like this to check for a specific version of a feature:

```
#if defined __cpp_constexpr && __cpp_constexpr >= 201304
```

Reply:
Indeed, but that first exemple would also trigger the `undefined warning` you are talking about, as it would not be defined anyway for c++98 ;)
I've always used a compiler more recent (with latest c++ features) than the code I produce, so I never tested for a compiler that don't support it (as it wouldn't compile my code anyway).

In general, I think the most robust way of handling lib features test is having both tests, even if there isn't (at the moment) more than one version. That way it's ready for a new version and the reader is aware this was tested for the defined version.
eg.
```
#ifdef __cpp_lib_map_try_emplace
#  if __cpp_lib_map_try_emplace >= 201411L
#  endif
#else
#endif
```

But in our case here, I replaced the #if with #ifdef and pushed it. Thanks for noticing this ;)